src/main/java/de/tum/in/www1/artemis/service/programming/ProgrammingExerciseGradingService.java

```diff
@@ -822,57 +812,53 @@ private double calculateSuccessfulTestPoints(final ProgrammingExercise programmi
-
         if (Double.isNaN(points)) {
-            cappedPoints = 0;
-        }
-        else if (points > maxPoints) {
-            cappedPoints = maxPoints;
+            return 0;
         }
 
-        return cappedPoints;
+        double maxPoints = programmingExercise.getMaxPoints() + Objects.requireNonNullElse(programmingExercise.getBonusPoints(), 0D);
+
+        return Math.min(points, maxPoints);
     }
 
     /**
      * Updates the feedback corresponding to the test case with the given credits.
-     * @param result which should be updated.
+     *
      * @param testCase the feedback that should be updated corresponds to.
-     * @param credits that should be set in the feedback.
+     * @param result   which should be updated.
+     * @param credits  that should be set in the feedback.
      */
-    private void setCreditsForTestCaseFeedback(final Result result, final ProgrammingExerciseTestCase testCase, double credits) {
+    private void setCreditsForTestCaseFeedback(final ProgrammingExerciseTestCase testCase, final Result result, double credits) {
         // We need to compare testcases ignoring the case, because the testcaseRepository is case-insensitive
         result.getFeedbacks().stream().filter(fb -> FeedbackType.AUTOMATIC.equals(fb.getType()) && fb.getText().equalsIgnoreCase(testCase.getTestName())).findFirst()
                 .ifPresent(feedback -> feedback.setCredits(credits));
     }
 
     /**
      * Calculates the points that should be awarded for a successful test case.
-     * @param result used to determine if the calculation is performed for the solution.
-     * @param programmingExercise the result belongs to.
-     * @param test for which the points should be calculated.
-     * @param totalTestCaseCount in the given exercise.
-     * @param weightSum of all test cases in the exercise.
+     *
+     * @param test              for which the points should be calculated.
      * @return the points which should be awarded for successfully completing the test case.
      */
-    private double calculatePointsForTestCase(final Result result, final ProgrammingExercise programmingExercise, final ProgrammingExerciseTestCase test, int totalTestCaseCount,
-            double weightSum) {
-        final boolean isWeightSumZero = Precision.equals(weightSum, 0, 1E-8);
+    private double calculatePointsForTestCase(final ProgrammingExerciseTestCase test, ScoreCalculationData scoreCalculationData) {
+        final int totalTestCaseCount = scoreCalculationData.testCases().size();
+
+        final boolean isWeightSumZero = Precision.equals(scoreCalculationData.weightSum(), 0, 1E-8);
         final double testPoints;
+        double exerciseMaxPoints = scoreCalculationData.exercise().getMaxPoints();
 
         // A weight-sum of zero would let the solution show an error to the instructor as the solution score must be
         // 100% of all reachable points. To prevent this, we weigh all test cases equally in such a case.
```

Comment:
This sounds to me like a hack especially for PGDP. I don't see why this is needed, let it show a warning that the solution did not got 100%. The instructor then knows that this is expected and can ignore that.

But In case such a configuration was not done intentionally, this warning should definitly be shown.

If you want, you can also just add a TODO comment and do this in a follow-up / pass it to me. If I remember this correctly, Artemis currently does not even allow to save such a configuration, which definitely should be allowed.

Reply:
> This sounds to me like a hack especially for PGDP.

We use such configurations, too, where the automatic tests should just help students check that the functionality works but the actual grading is done manually based on code style, structure and/or efficiency of the implementation.

> If I remember this correctly, Artemis currently does not even allow to save such a configuration, which definitely should be allowed.

It is allowed.

If it always shows 0% for the solution and a warning in such cases then it would be harder for the instructor to see if
- the solution does indeed pass all tests and the 0% is expected, or
- there are actually >=1 test failures, which in all cases still results in 0%, but of course is an issue.

So any change here should be probably accompanied by an update to the client code that changes how the result is shown for the solution in such cases:
- 0%, but all tests pass: okay,
- 0%, but there are test failures: show a warning.

I guess then the special case is needed there for the solution instead?

⇒  +1 for moving that to a future PR; maybe not even as a ToDo comment, but a proper GH issue

Reply:
Thanks for the insight, good point in distinguishing between "All Tests passed but 0%" and "0% because not all tests passed". Also, what's about the case "All tests passed, but there are SCA issue which should deduct points, but cannot deduct points since the solution is at 0%"?

> It is allowed.

Yes, but currently only in your case, if manual correction is enabled. For programming exercises with manual correction, it's not allowed. However, we also had use cases for this in the past.

https://github.com/ls1intum/Artemis/blob/develop/src/main/java/de/tum/in/www1/artemis/service/programming/ProgrammingExerciseTestCaseService.java#L139

https://github.com/ls1intum/Artemis/blob/develop/src/main/webapp/app/exercises/programming/manage/grading/programming-exercise-configure-grading.component.ts#L818

I'll look further into this in the future.

Reply:
#6093